rpcs3/Crypto/unpkg.cpp

```diff
@@ -744,180 +747,258 @@ bool package_reader::extract_data(atomic_t<double>& sync)
+	std::memcpy(entries.data(), m_bufs.back().get(), entries.size() * sizeof(PKGEntry));
 
+	// Create directories first
 	for (const auto& entry : entries)
 	{
-		if (entry.name_size > 256)
+		if (entry.name_size > PKG_MAX_FILENAME_SIZE)
 		{
 			num_failures++;
 			pkg_log.error("PKG name size is too big (0x%x)", entry.name_size);
-			continue;
+			break;
 		}
 
-		const bool is_psp = (entry.type & PKG_FILE_ENTRY_PSP) != 0u;
+		switch (const u8 entry_type = entry.type & 0xff)
+		{
+		case PKG_FILE_ENTRY_FOLDER:
+		case 0x12:
+		{
+			const bool is_psp = (entry.type & PKG_FILE_ENTRY_PSP) != 0u;
+			decrypt(entry.name_offset, entry.name_size, is_psp ? PKG_AES_KEY2 : m_dec_key.data());
 
-		decrypt(entry.name_offset, entry.name_size, is_psp ? PKG_AES_KEY2 : m_dec_key.data());
+			const std::string name{reinterpret_cast<char*>(m_bufs.back().get()), entry.name_size};
+			const std::string path = dir + vfs::escape(name);
 
-		const std::string name{reinterpret_cast<char*>(m_buf.get()), entry.name_size};
-		const std::string path = dir + vfs::escape(name);
+			const bool log_error = entry.pad || (entry.type & ~PKG_FILE_ENTRY_KNOWN_BITS);
 
-		const bool log_error = entry.pad || (entry.type & ~PKG_FILE_ENTRY_KNOWN_BITS);
-		(log_error ? pkg_log.error : pkg_log.notice)("Entry 0x%08x: %s (pad=0x%x)", entry.type, name, entry.pad);
+			(log_error ? pkg_log.error : pkg_log.notice)("Entry 0x%08x: %s (pad=0x%x)", entry.type, name, entry.pad);
 
-		switch (const u8 entry_type = entry.type & 0xff)
-		{
-		case PKG_FILE_ENTRY_NPDRM:
-		case PKG_FILE_ENTRY_NPDRMEDAT:
-		case PKG_FILE_ENTRY_SDAT:
-		case PKG_FILE_ENTRY_REGULAR:
-		case PKG_FILE_ENTRY_UNK0:
-		case PKG_FILE_ENTRY_UNK1:
-		case 0xe:
-		case 0x10:
-		case 0x11:
-		case 0x13:
-		case 0x14:
-		case 0x15:
-		case 0x16:
-		case 0x18:
-		case 0x19:
-		{
-			const bool did_overwrite = fs::is_file(path);
-
-			if (did_overwrite && !(entry.type & PKG_FILE_ENTRY_OVERWRITE))
-			{
-				pkg_log.notice("Didn't overwrite %s", path);
+			if (fs::is_dir(path))
+			{
+				pkg_log.warning("Reused existing directory %s", path);
+			}
+			else if (fs::create_path(path))
+			{
+				pkg_log.notice("Created directory %s", path);
+			}
+			else
+			{
+				num_failures++;
+				pkg_log.error("Failed to create directory %s", path);
 				break;
 			}
 
-			const bool is_buffered = entry_type == PKG_FILE_ENTRY_SDAT;
+			break;
+		}
+		default:
+		{
+			continue;
+		}
+		}
+	}
+
+	if (num_failures != 0)
+	{
+		if (was_null)
+		{
+			fs::remove_all(dir, true);
+		}
+
+		pkg_log.error("Package installation failed: %s", dir);
+		return false;
+	}
+
+	atomic_t<usz> g_entry = 0;
+	atomic_t<usz> g_thread_index = 0;
+
+	m_memory_file = m_file.map_to_memory(0, m_file.size());
+
+	if (m_memory_file)
+	{
+		m_bufs.resize(std::min<usz>(utils::get_thread_count(), entries.size()));
+	}
+
+	auto install_func = [&]()
+	{
+		const thread_key thread_data_key{g_thread_index++};
```

Comment:
I don't see a reason to have this confusing struct at all if you're just passing a usz anyway.